src/OpenApi/src/Services/OpenApiDocumentService.cs

```diff
@@ -1,23 +1,74 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.AspNetCore.Mvc.ApiExplorer;
+using Microsoft.Extensions.DependencyInjection;
 using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Options;
 using Microsoft.OpenApi.Models;
 
 namespace Microsoft.AspNetCore.OpenApi;
 
-internal sealed class OpenApiDocumentService(IHostEnvironment hostEnvironment)
+internal class OpenApiDocumentService(
+    [ServiceKey] string documentName,
+    IApiDescriptionGroupCollectionProvider apiDescriptionGroupCollectionProvider,
+    IHostEnvironment hostEnvironment,
+    IOptionsSnapshot<OpenApiOptions> optionsSnapshot)
 {
+    private readonly OpenApiOptions _options = optionsSnapshot.Get(documentName);
+
     public Task<OpenApiDocument> GetOpenApiDocumentAsync()
     {
         var document = new OpenApiDocument
         {
-            Info = new OpenApiInfo
-            {
-                Title = hostEnvironment.ApplicationName,
-                Version = OpenApiConstants.DefaultOpenApiVersion
-            }
+            Info = GetOpenApiInfo(),
+            Paths = GetOpenApiPaths()
         };
         return Task.FromResult(document);
     }
+
+    internal OpenApiInfo GetOpenApiInfo()
+    {
+        return new OpenApiInfo
+        {
+            Title = $"{hostEnvironment.ApplicationName} | {documentName}",
+            Version = OpenApiConstants.DefaultOpenApiVersion
+        };
+    }
+
+    /// <summary>
+    /// Gets the OpenApiPaths for the document based on the ApiDescriptions.
+    /// </summary>
+    /// <remarks>
+    /// At this point in the construction of the OpenAPI document, we run
+    /// each API description through the `ShouldInclude` delegate defined in
+    /// the object to support filtering each
+    /// description instance into its appropriate document.
+    /// </remarks>
+    internal OpenApiPaths GetOpenApiPaths()
+    {
+        var descriptionsByPath = apiDescriptionGroupCollectionProvider.ApiDescriptionGroups.Items
+            .SelectMany(group => group.Items)
```

Comment:
A group is what we call a document?  Do items know their groups?

Reply:
Yes, with the default inclusion predicate the group name maps to the document name. It's possible to customize the `ShouldInclude` function to collate endpoints into a document based on other heuristics so it's not _always_ the case.

And, yes, [items know their groups](https://source.dot.net/#Microsoft.AspNetCore.Mvc.Abstractions/ApiExplorer/ApiDescription.cs,23).